validations/src/validations.rs

```diff
@@ -150,30 +155,162 @@ pub fn validate_rad_request(rad_request: &RADRequest) -> Result<(), failure::Err
+    fn update(&mut self, pos: usize) {
+        self.values[pos] += 1;
+        if self.values[pos] > self.max_val {
+            self.max_pos = pos;
+            self.max_val = self.values[pos];
+        }
+    }
+}
+
+fn update_liars(liars: &mut Vec<bool>, item: RadonTypes, condition: bool) -> Option<RadonTypes> {
+    if condition {
+        liars.push(false);
+        Some(item)
+    } else {
+        liars.push(true);
+        None
+    }
+}
+
+pub fn tally_precondition_clause(
+    reveals: Vec<RadonReport<RadonTypes>>,
+    non_error_min: f64,
+) -> Result<(Vec<RadonTypes>, Vec<bool>), RadError> {
+    let reveals_len = reveals.len() as f64;
+
+    let mut counter = Counter::default();
+    for reveal in &reveals {
+        match &reveal.result {
+            Err(_e) => {}
+            Ok(rad_types) => match rad_types {
+                RadonTypes::Array(_) => counter.update(0),
+                RadonTypes::Boolean(_) => counter.update(1),
+                RadonTypes::Bytes(_) => counter.update(2),
+                RadonTypes::Float(_) => counter.update(3),
+                RadonTypes::Integer(_) => counter.update(4),
+                RadonTypes::Map(_) => counter.update(5),
+                RadonTypes::String(_) => counter.update(6),
+            },
+        }
+    }
+
+    let non_error_ratio = counter.max_val as f64 / reveals_len;
+
+    if non_error_ratio > non_error_min {
+        let mut liars = vec![];
+        let results = reveals
+            .into_iter()
+            .filter_map(|reveal| match reveal.into_inner() {
+                Err(_e) => {
+                    liars.push(true);
+                    None
+                }
+                Ok(rad_types) => match rad_types {
+                    RadonTypes::Array(_) => {
+                        update_liars(&mut liars, rad_types, counter.max_pos == 0)
+                    }
+                    RadonTypes::Boolean(_) => {
+                        update_liars(&mut liars, rad_types, counter.max_pos == 1)
+                    }
+                    RadonTypes::Bytes(_) => {
+                        update_liars(&mut liars, rad_types, counter.max_pos == 2)
+                    }
+                    RadonTypes::Float(_) => {
+                        update_liars(&mut liars, rad_types, counter.max_pos == 3)
+                    }
+                    RadonTypes::Integer(_) => {
+                        update_liars(&mut liars, rad_types, counter.max_pos == 4)
+                    }
+                    RadonTypes::Map(_) => update_liars(&mut liars, rad_types, counter.max_pos == 5),
+                    RadonTypes::String(_) => {
+                        update_liars(&mut liars, rad_types, counter.max_pos == 6)
+                    }
+                },
+            })
+            .collect();
+
+        Ok((results, liars))
+    } else {
+        let errors: Vec<RadError> = reveals
+            .into_iter()
+            .filter_map(|reveal| match reveal.into_inner() {
```

Comment:
This match implements [Result::ok](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok)

```suggestion
            .filter_map(|reveal| reveal.into_inner().ok())
```

Reply:
But it is inverse. Err -> Some, Ok -> None